h2/src/main/org/h2/schema/Schema.java

```diff
@@ -149,14 +149,25 @@ public void removeChildrenAndResources(Session session) {
         removeChildrenFromMap(session, constraints);
         // There can be dependencies between tables e.g. using computed columns,
         // so we might need to loop over them multiple times.
+        boolean modified = true;
         while (!tablesAndViews.isEmpty()) {
+            boolean newModified = false;
             for (Table obj : tablesAndViews.values()) {
                 // Database.removeSchemaObject() removes the object from
                 // the map too, but it is safe for ConcurrentHashMap.
-                if (database.getDependentTable(obj, obj) == null) {
+                Table dependentTable = database.getDependentTable(obj, obj);
+                if (dependentTable == null) {
                     database.removeSchemaObject(session, obj);
+                    newModified = true;
+                } else if (dependentTable.getSchema() != this) {
+                    throw DbException.get(ErrorCode.CANNOT_DROP_2, obj.getSQL(false), dependentTable.getSQL(false));
+                } else if (!modified) {
+                    dependentTable.removeColumnExpressionsDependencies(session);
+                    dependentTable.setModified();
+                    database.updateMeta(session, dependentTable);
```

Comment:
this modified/newModified stuff is rather confusing - what is going on here?

Reply:
If no table was deleted in the previous iteration it means that there is a circular dependency and this fallback code is activated. It clears the suspicious properties of all columns.

Reply:
ah, I see, thanks

Reply:
Hypothetically such properties can be dropped immediately, but (a) I'm not sure that it is safe without `updateMeta()` and (b) I don't want to call `updateMeta()` from here when it isn't really required.

This solution is temporary with better-than-nothing approach.

I think that we need to redesign dependencies. For example, the `Database.getDependentTable()` is very slow on databases with many tables.

Reply:
If Database.getDependentTable() is slow, we can just build a dependency map once at the beginning of this routine, instead of doing it inside the loop.

And we can do a two pass thing, where we 
(1 loop) strip all the tables of foreign keys, constraints,  reset views to "select 1", etc
(2 loop) and then remove the tables.

Reply:
`DROP TABLE a, b, …` is also problematic. And it isn't fixed here. But at least it does not go into infinite loop, it simply throws exceptions on circular dependencies. Most likely other `DROP` commands have some issues too.

I think that such logic with resolution of dependencies should not be placed here. It should be reusable for all DDL commands.